<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px" />

# Python Ladder Challenge - Notebook 2

_Author:_ Tim Book

# Climb the Ladder!
Our class moves quickly! Sometimes, it feels like we make leaps in logic that are a bit too big. Learn Python _slowly_, by doing many, many examples. Problems in this notebook start out easy and progressively get harder, so that the next rung of the Python ladder is always within reach.

# Section IV: Conditionals and Loops

34) Print out the numbers 1 through 10.

In [3]:
number_list = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
print(number_list[0:11])

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


35) Print out the odd numbers between 1 and 20 by using the third paramter to the `range()` function.

In [5]:
list(range(1, 20, 2))

[1, 3, 5, 7, 9, 11, 13, 15, 17, 19]

36) Print out the odd numbers between 1 and 20 by checking if the number is even before printing.

In [9]:
for num in range(1,21):
    if num % 2 == 1:
        print(num)


1
3
5
7
9
11
13
15
17
19


37) What is the sum of the first 100 integers? Hint: Create a variable `total` equal to zero, and increment `total` inside a loop.

In [11]:
total = 0 

for n in range(101):
    total += n
print(total)

5050


38) For each friend in your `friends` list, print `'PERSON IS AWESOME!'` in all caps, where `PERSON` is of course replaced with your friend's name.

In [13]:
friends = ["Alice", "Bob", "Charlie", "Debbie"]

In [17]:
for friend in friends:
    print(friend.upper(), 'IS AWESOME')


ALICE IS AWESOME
BOB IS AWESOME
CHARLIE IS AWESOME
DEBBIE IS AWESOME


39) Similar to the last problem, print `'PERSON IS AWESOME!'` for each friend in your friends list _only if their name ends in a vowel_. Otherwise, print `"person is ok..."` (all lowercase). 

In [21]:
for friend in friends:
    vowel = ['a', 'e', 'i', 'o', 'u']
    if friends.endswith(vowel):
        print(friend.upper(), 'IS AWESOME')
    else:
        print(friend, 'is ok....')

AttributeError: 'list' object has no attribute 'endswith'

40) For each number between 1 and 30 (inclusive), if a number is divisible by 3, print `'Fizz'`. Otherwise, just print the number.

In [22]:
for num in range(1,31):
    if num % 3 == 0:
        print('Fizz')
    else:
        print(num)

1
2
Fizz
4
5
Fizz
7
8
Fizz
10
11
Fizz
13
14
Fizz
16
17
Fizz
19
20
Fizz
22
23
Fizz
25
26
Fizz
28
29
Fizz


41) For each number between 1 and 30 (inclusive), if a number is divisible by 5, print `'Buzz'`. Otherwise, just print the number.

In [23]:
for num in range(1,31):
    if num % 5 == 0:
        print('Fizz')
    else:
        print(num)

1
2
3
4
Fizz
6
7
8
9
Fizz
11
12
13
14
Fizz
16
17
18
19
Fizz
21
22
23
24
Fizz
26
27
28
29
Fizz


42) **FIZZBUZZ!** For each number between 1 and 30 (inclusive),
* if a number is divisible by 3, print `'Fizz'`
* if a number is divisible by 5, print `'Buzz'`
* if a number is divisible by both 3 and 5, instead print `'FizzBuzz'`,
* otherwise, just print the number

This problem is the famous "Fizzbuzz" problem - and is a very common interview coding challenge! If you can do this without help, you're making awesome progress! Well done!

In [24]:
for num in range(1,31):
    if num % 3 == 0 and num % 5 == 0:
        print('FizzBuzz')
    elif num % 3 == 0:
        print('Fizz')
    elif num % 5 == 0:
        print('Buzz')
    else:
        print(num)

1
2
Fizz
4
Buzz
Fizz
7
8
Fizz
Buzz
11
Fizz
13
14
FizzBuzz
16
17
Fizz
19
Buzz
Fizz
22
23
Fizz
Buzz
26
Fizz
28
29
FizzBuzz


43) Define some string `sentence`. If you can't think of a good one, feel free to use `fact` from the previous notebook.

In [31]:
sentence = 'Python Programming is Fun'
print(sentence)

Python Programming is Fun


44) Count the number of vowels in `sentence` (no, y is not a vowel).
* _Hint:_ What happens if you loop through a string?

In [36]:
vowel_count = 0

for vowels in sentence:
    if vowels in 'aeiou':
        vowel_count += 1
print(vowel_count)   


6


45) Define a variable `sentence_rm` that is equal to `sentence` with all vowels removed. 
* _Hint:_ This is a tough one. Start out with an empty string and concatenate all of the consonants to it inside a loop.

46) Create a variable `fav_number` equal to whatever is your favorite integer. Write a loop to determine if `fav_number` is prime. Recall that a prime number is any number whose only divisors are 1 and itself. For example, 6 is not prime since it's divisible by 1, 2, 3, and 6. But 7 _is_ prime since it's only divisible by 1 and 7.

47) A palindromic number reads the same both ways. For example, 1234321 is a palindrome. Write an `if` statement to test whether or not a number is palindromic.

48) Find the largest palindrome made from the product of two two-digit numbers.
* _Tip:_ Do not worry about the "efficiency" of your answer! The easiest answer is very inefficient.

# Section V: List Comprehensions

**Disclaimer:** Many (maybe all) of these exercises can be done without a list comprehension. However, listcomps are often the easiest way to solve a problem like this. Listcomps are also optimized in Python, meaning they are also often the _fastest_ way to solve a problem. In fact, **every problem in this section can be solved in one line of code!**

---
**IMPORTANT NOTE: Consider all ranges in this section to be *inclusive*.**

49) Create a list of the numbers 1 through 10.

In [38]:
n_list = [num for num in range(1, 11)]
n_list

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

50) Create a list of the squares of the numbers between 1 and 10.

In [42]:
n_list = [num ** 0.5 for num in range(1,11)]
print(n_list)

[1.0, 1.4142135623730951, 1.7320508075688772, 2.0, 2.23606797749979, 2.449489742783178, 2.6457513110645907, 2.8284271247461903, 3.0, 3.1622776601683795]


51) Create a list of all the even numbers between 18 and 47.

In [43]:
list_even = [num for num in range(18,48) if num % 2 ==0]
list_even

[18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46]

52) Create a list of the squares of the even numbers between 1 and 20.

In [45]:
list_squares = [num ** 0.5 for num in range(1,21) if num % 2 == 0]
list_squares

[1.4142135623730951,
 2.0,
 2.449489742783178,
 2.8284271247461903,
 3.1622776601683795,
 3.4641016151377544,
 3.7416573867739413,
 4.0,
 4.242640687119285,
 4.47213595499958]

53) What is the sum of the cubes of the numbers between 7 and 37?

54) Create a list of all of the numbers divisible by either 3 or 5 (or both) between 1 and 30.

In [50]:
list_fizzbuzz = [num for num in range(1,31) if num % 3 == 0 and num % 5 == 0]
list_fizzbuzz

[15, 30]

55) Redefine this list to be all uppercase.

In [69]:
friends = ['Alice', 'Bob', 'Charlie', 'Derek']

In [57]:
upper_friends = []
for friend in friends:
    upper_friends.append(friend.upper())
upper_friends

['ALICE', 'BOB', 'CHARLIE', 'DEBBIE']

56) Remove all elements from `friends` whose names do not end in a vowel.

In [66]:
for friend in friends:
    vowel = 'aeiou'
    if friend.endswith(vowel):
        friends.remove
friends

['Alice', 'Bob', 'Charlie', 'Debbie']

57) Take the `sentence` you used earlier and remove the vowels again. This time, using a one-line list comprehension.

* _Hint:_ You'll find the `.join()` method to be useful here.

58) You need to access Excel files that have names like `Jan 5.xlsx` or `Mar 15.xlsx`. Create a list of January 1 through January 31 like this. That is, create the list of: `['Jan 1.xlsx', 'Jan 2.xlsx', ... , 'Jan 31.xlsx']`.

59) Below is a list of amounts in euros. Create a new list, `dollars`, which is these amounts converted to U.S. dollars. There are 1.1 dollars per euro.

In [ ]:
euros = [4.50, 6.70, 3.25, 9.99, 12.75, 0.35]

60) Below is a list of heights in inches. In order to ride a roller coaster, you must be at least 5 feet tall (60 inches). Filter this list to be _only_ those heights tall enough to ride.

In [ ]:
heights = [71, 48, 55, 65, 68, 60, 58, 53]

61) Repeat the above exercise, except replace the lower heights with a `None` instead of dropping them.
* _Hint:_ Python has an in-line `if` statement, sometimes called a _ternary operator_ that you might find useful here. For example:

```
print("Tall enough" if 72 > 60 else "Too short")
print("Tall enough" if 52 > 60 else "Too short")
```

62) Here are more heights below. However, this time, the information is contained in a dictionary. Create a list of the names _and ONLY the names_ of the people who are tall enough to ride the roller coaster.

* _Hint:_ How can you loop through a dictionary?

In [ ]:
people = {
    'Aaron': 58,
    'Barbara': 66,
    'Clarence': 62,
    'Donovan': 55,
    'Erika': 70,
    'Fernando': 72
}

63) Below we have some more data on our classmates. This time, the dictionary values are test scores. A student's final grade is their _maximum_ score on these three tests. Create a list of the students' final grades.

In [70]:
people = {
    'Aaron': [87, 52, 78],
    'Barbara': [92, 79, 85],
    'Clarence': [42, 68, 55],
    'Donovan': [95, 100, 87],
    'Erika': [62, 88, 47],
    'Fernando': [84, 99, 0]
}

64) Repeat the above problem, except create a _dictionary_ of the final scores, where the dictionary keys are the names, and the dictionary values are the final grade.

* _Hint:_ You can do **dictionary comprehension** to solve this problem! Yes, it exists!

65) Using the two lists defined below, create the following resulting list:

`['AZ', 'BY', 'CX', 'DW', 'EV']`

* _Hint:_ Check out the `zip()` function.

In [ ]:
letters_a = ['A', 'B', 'C', 'D', 'E']
letters_z = ['Z', 'Y', 'X', 'W', 'V']

66) Using `letters_a` defined in the previous problem, create the following list:

```
[
    'A is letter 1 of the alphabet',
    'B is letter 2 of the alphabet',
    ...
]
```

* _Hint:_ Checkout the `enumerate()` function.

# Section VI: Libraries, imports and some more math!

67) Import the `math` library

68) The `math` library has a lot of cool stuff in it. For this problem, evaluate the following math expressions:

a) $\sin{\frac{\pi}{2}}$

b) $\ln{\sqrt{e}}$
* _Hint:_ A logarithm with base $e$ is represented by just `log` in `math`. In fact, throughout the course, $\log$ with no subscript will always denote the natural log (ie, log base $e$).

c) $e^{3!}$

69) Two popular math functions are the "round up" and "round down" functions. In computer science, we call them the "ceiling" and "floor" functions. They're used and denoted as follows:

* `ceil(3.8)` $= \lceil 3.8 \rceil = 4$
* `ceil(3.2)` $= \lceil 3.2 \rceil = 4$
* `floor(3.8)` $= \lfloor 3.8 \rfloor = 3$
* `floor(3.2)` $= \lfloor 3.2 \rfloor = 3$

We _only_ need these two functions out of the `math` library. Use an `import` statement to import _only_ these two functions and verify the above examples are true.

70) We can also _alias_ library names if they're too long and we will be using them often. As we will do many times in this course, import the following three libraries:

* Import the `numpy` library, aliasing it as `np`
* Import the `pandas` library, aliasing it as `pd`
* Import the `matplotlib.pyplot` library and submodule, aliasing it as `plt`

71) You can also import everything from a library all at once without needing to prefix it:

`from math import *`

For this question your task is: _never, **ever**_ do this.

Why?
1. It "pollutes" the namespace with variable names that could easily be overwritten on accident.
    - For example, after the import we could accidentally define `sin = 'without'` then be surprised when `sin(0.5)` no longer works!
2. It does not show the reader which module a given function came from. 
    - For example, we see above that `ceil` came from the `math` library since it was noted in the import.
    - If we use * is much more difficult to determine which module `ceil` is from.